stellargraph/core/graph.py

```diff
@@ -279,6 +274,25 @@ def __init__(
             weight_column=edge_weight_column,
         )
 
+        edge_node_set = set(self._edges.sources)
+        edge_node_set.update(self._edges.targets)
+
+        if nodes is None and edges is not None:
+            nodes = pd.DataFrame([], index=edge_node_set)
+        elif nodes is None:
+            nodes = {}
+
+        self._nodes = convert.convert_nodes(
+            nodes, name="nodes", default_type=node_type_default, dtype=dtype,
+        )
+
+        node_set = set(self._nodes.ids.pandas_index)
+        if not edge_node_set.issubset(node_set):
+            raise ValueError(
+                f"Found nodes referenced in `edges` not contained in `nodes`. Either add these nodes "
```

Comment:
I think this error message can be tweaked to be more specific, e.g. using `comma_sep`:

```suggestion
                f"edges: expected all source and target node IDs to be contained in `nodes`, found some missing: {comma_sep(edge_node_set - node_set)}"
```

(Or however we end up doing the check.)